wally_the_cartographer/cmd/root.go

```diff
@@ -1,55 +1,459 @@
+			WorkQueue:    []WorkItem{},
+			PullRequests: make(map[int]*PullRequestState),
+		}
+	}
+	repositoryState := state.Repositories[repoName]
+
+	// Handle all WorkItems.
+	for len(repositoryState.WorkQueue) != 0 {
+		workItem := repositoryState.WorkQueue[0]
+		// TODO(hrapp): Only do this if no further comments have been made and last_sha_seen has not changed.
+		// TODO(hrapp): Check if the PR is still open.
+		pullRequest, _, err := client.PullRequests.Get(ctx, *repo.Owner.Login, *repo.Name, workItem.Pr)
+		if err != nil {
+			return err
+		}
+
+		var logOutput bytes.Buffer
+		logWriter := io.MultiWriter(&logOutput, os.Stdout)
+		done, err := handleWorkItem(ctx, repositoryState.PullRequests[workItem.Pr], client, pullRequest, repo, datadir, logWriter)
+		if err != nil {
+			_ = postCommentToPr(ctx, client, repo, pullRequest.GetNumber(),
+				fmt.Sprintf("Error: %v\n\nLog:\n~~~\n%s\n~~~", err, logOutput.String()))
+			return err
+		}
+		if err := WriteState(state, datadir); err != nil {
+			return err
+		}
+		if !done {
+			// Did some work, but is not done. We have to check again next run of the tool.
+			break
+		}
+		repositoryState.WorkQueue = repositoryState.WorkQueue[1:]
+	}
+
+	// Look for new comments on PullRequests..
+	pullRequests, _, err := client.PullRequests.List(ctx, *repo.Owner.Login, *repo.Name, &github.PullRequestListOptions{})
+	if err != nil {
+		return err
+	}
+
+	mergeRegex, err := regexp.Compile("(?m:^@" + userName + `\s*merge$)`)
+	if err != nil {
+		log.Fatalf("Could not compile regex: %v", err)
+	}
+
+	for _, pullRequest := range pullRequests {
+		prNum := pullRequest.GetNumber()
+		p := repositoryState.PullRequests[prNum]
+		if p == nil {
+			p = &PullRequestState{}
+			repositoryState.PullRequests[prNum] = p
+		}
+		comments, _, err := client.Issues.ListComments(ctx, *repo.Owner.Login, *repo.Name, prNum, &github.IssueListCommentsOptions{})
+		if err != nil {
+			return err
+		}
+
+		for _, comment := range comments[p.NumComments:] {
+			if len(mergeRegex.FindString(*comment.Body)) == 0 {
+				continue
+			}
+
+			membership, _, err := client.Organizations.GetTeamMembership(ctx, teamID, *comment.User.Login)
+			if err != nil {
+				return err
+			}
+
+			if *membership.State != "active" {
+				log.Printf("User %s is not in our team. Ignoring merge request.", *comment.User.Login)
+				continue
+			}
+
+			repositoryState.WorkQueue = append(repositoryState.WorkQueue, WorkItem{Repo: repoName, Pr: prNum})
+			_ = postCommentToPr(ctx, client, repo, prNum,
+				fmt.Sprintf("Merge requested by authorized user %s. Merge queue now has a length of %d.",
+					*comment.User.Login,
+					len(repositoryState.WorkQueue)))
+			// We ignore all further comments on this PR and will never look at them.
+			break
+		}
+		p.NumComments = len(comments)
+		p.LastSHASeen = *pullRequest.Head.SHA
+	}
+	return nil
+}
+
+// RootCmd is a cobra concept and requires a comment to shut up the linter.
```

Comment:
Is there nothing more useful that you can say about RootCmd?

Reply:
no. I really don't quite understand how Cobra works.